### ISSUE TITLE
Concurrent API Requests

### ISSUE THREAD

**Issue by hamisor** [mgp25/Instagram-API#1584]
### Before submitting an issue make sure you have:
- [x] [Searched](https://github.com/mgp25/Instagram-API/search?type=Issues) the bugtracker for similar issues including **closed** ones
- [x] [Read the FAQ](https://github.com/mgp25/Instagram-API/wiki/FAQ)
- [x] [Read the wiki](https://github.com/mgp25/Instagram-API/wiki)
- [x] [Reviewed the examples](https://github.com/mgp25/Instagram-API/tree/master/examples)
- [x] [Installed the api using ``composer``](https://github.com/mgp25/Instagram-API#installation)

### Purpose of your issue?
- [ ] Bug report (encountered problems/errors)
- [x] Feature request (request for new functionality)
- [ ] Question
- [ ] Other

### Describe your issue

Hi,
Firstly, thank you for this amazing library. The feature I would like is the following: in order best simulate the IG API I would like to make particular combinations of concurrent requests, just like the mobile app does. 

If implemented correctly this should have no detrimental effects, as the API already implements a leaky-bucket algorithm on a per-endpoint level (not at the account-level). This means that concurrent API calls are ok as long as they are not calling the same endpoint. 

This could be achieved using guzzle concurrent requests (more info at http://docs.guzzlephp.org/en/stable/quickstart.html#concurrent-requests). The library would have to be modified such that instead of returning data from IG, it would instead return guzzle promises. 

I would like to do this by implementing a new endpoint, for example:

```php

use GuzzleHttp\Promise;

class People extends RequestCollection
{
	function getUserProfile($igUserId)
	{
		// Initiate each request but do not block
		$promises = [
			'info'    => $this->getInfoById_Async($igUserId),
			'user_feed'  => $this->getUserFeed_Async($igUserId),
			'story_feed'  => $this->getUserStoryFeed_Async($igUserId),
			'friendship'  => $this->getFriendship_Async($igUserId)
		];

		// Wait on all of the requests to complete. Throws a ConnectException
		// if any of the requests fail
		$results = Promise\unwrap($promises);

		// Wait for the requests to complete, even if some of them fail
		$results = Promise\settle($promises)->wait();

		return $results;
	}
}

```

```php
 $this->ig->people->getUserProfile($igUserId);
```
which would asynchronously call the following endpoints under the hood:

- getInfoById()
- getUserFeed()
- getUserStoryFeed()
- getFriendship()

This would closely mimic the app-behaviour when clicking on a username.

I would like to fork the library and implement this as a feature, but I would like to get your feedback on the architectural/implementation details of my plan; namely the call stack looks roughly like the following:

- Request->getResponse()
  - Request->getHttpResponse()
    - $this->_parent->client->api() 
      - Client->_apiRequest()
        - _guzzleRequest()
          - _guzzleClient->send()

I would like to `duplicate` each function with an asynchronous partner-function (e.g. _apiRequest and apiRequestAsync). The end result would be that each endpoint can have an async partner (e.g. getInfoById and getInfoByIdAsync). However, the async partner-functions should be private in order to avoid api consumers getting blocked. Instead, they should be wrapped into a safe function that will also manage concurrency (see the example of `getUserProfile()` above). 

Thanks

**Comment by valga** [comment]
> I would like to duplicate each function with an asynchronous partner-function (e.g. _apiRequest and apiRequestAsync).

Proper solution is to make the core methods async, so the request will always return a `Promise` object.

> However, the async partner-functions should be private in order to avoid api consumers getting blocked.

Async methods in request collections must be public. So you should end up with something like this:
```php
/**
  * @return Promise
 */
public function callSomeEndpointAsync($param)
{
    return $this->ig->request('/endpoint')
        ->addParam('param', $param)
        ->getResponse(new SomeResponse());
}

/**
  * @throws InstagramException
  *
  * @return SomeResponse
 */
public function callSomeEndpoint($param)
{
    return $this->callSomeEndpointAsync($param)->wait();
}
```

Notice the sync method is now a partner to the async one.

> Instead, they should be wrapped into a safe function that will also manage concurrency (see the example of getUserProfile() above).

It should be done just for `login()` flows.

To conclude, it's a good idea, but you should not underestimate the complexity of this task. Just look at the upload methods, for example the video ones. 

Probably you can just convert the core to async, and that will work for your goal. I thought of this (I mean async requests) while making Realtime&Push clients (because making sync requests in async environment is a pretty bad idea), but I discarded it, because it way too complex to rewrite everything to async calls.

**Comment by hamisor** [comment]
That's great, I really like the sync method calling async->wait() 

Again, I will implement this by duplicating the core methods with async equivalents. If you guys like the code, then it will be easy to implement the sync functions. And everything can live side-by-side.

I will probably avoid trying to make uploads async, at least initially.

Thanks,

**Comment by SteveJobzniak** [comment]
Hi @hamisor. It's absolutely fantastic to see that a clearly intelligent and brilliant mind like yours is interested in helping out on the project! And I'm glad that you like the project!

However... anything we spend time coding must have a clear value to be worth doing it. The only value of async endpoint calls would be in your `getUserProfile()` scenario, to emulate the real app calling 4 endpoints simultaneously.

This moves into the territory of trying to _exactly_ emulate the real app, when you visit profiles, or when you tap certain tabs to see some stats that comes from multiple endpoints, etc.

What are the advantages of multiple concurrent async requests in those cases? Well, they would save half a second compared to doing all 4 (or whatever) endpoint calls sequentially.

But then we get into another question: Do we _need_ to do all 4 endpoint calls? And the answer is no. We don't. Instagram does not in any way verify that the user is calling all 4 endpoints. It's fine to call only a single one of the user-related endpoints, and to call a single endpoint repeatedly, etc.

So there's absolutely no reason for us to call all 4 endpoints, and therefore no reason for us to have asynchronous code to efficiently call 4 endpoints.

Furthermore, even from a perspective of "trying to perfectly emulate the app so that Instagram can't know we're an unofficial client", it still doesn't make any sense. Because asynchronous multi-calls do NOT make us look like the real app. In the real app, they can change headers and function parameters all the time (which we won't know in this API; so they'll be able to detect us easily by just publishing a new version of their app, with a new param that we don't have, and voila, we're detected). And even worse: In the real app, the app constantly collects behavior-data ("graph"/analytics) about what you are doing in the app, etc. They use it for advertising and for proving that you are a human. The real app constantly sends that "graph"-data to an endpoint that we have not implemented _at all_ because it would be _way_ too much work.

Lastly, as valga said: Do not underestimate the amount of work to make the library asynchronous. Even valga, an absolutely brilliant and insanely good programmer who is very familiar with all of our code, and who wrote the async Push/Realtime clients, decided not to do it. **That** should give you a hint. ;-)

Furthermore: Users would gain no benefit by having async code. They would just get more complicated function-calls that return promises instead of final `ResponseInterface` objects. And the API would get bloated with duplicate versions of each API function (one async, one non-async).

To avoid that bloating, I think I would suggest a `InstagramAPI::setAsync(true/false)` instead of doing it on a per-function level. All functions already call `InstagramAPI::request()` to do their call, and then they call `Request::getResponse()`. So just rewrite the `Request::getResponse()` function, to make it return a Promise if `_parent->useAsync` is true, otherwise if false just connect and return the actual response object instantly as usual. It's then up to the caller (of `getResponse()`) to know how to deal with promises. So, the library user would get Promises from most functions if they've enabled Async mode. And Request objects themselves must also have some kinda `setAllowAsync(false)` (default to true), which will be used inside inner functions such as the video-uploader to guarantee that those ignore the global `$ig->useAsync` value and _always_ resolve and return the response instantly, so that their tight upload-loops can resolve themselves constantly without worrying about promises.

Under that kind of algorithm, the library would be unharmed by async capabilities. Because almost all functions just call a single endpoint and return its value instantly. Any functions that do multi-endpoint work can simply set `setAllowAsync(false)` on each of their own request objects, so that ALL of them have to be resolved instantly, one by one, before returning the result.

We can then easily add stuff like `getUserProfile($userId) { return [ 'info' => $this->getUserInfoById($userId), 'etc...' => ... ]; }`.

If the library is in non-async mode, the user would get a return value an array with all of the API responses in it.

If the library is in async mode, the user would get a return value as an array of promises.

And for single-endpoint functions (that don't call multiple ones), the user would get either a Promise, or a full ResponseInterface object.

So if the user is in non-async mode, they would just get simple ResponseInterface objects or arrays of ResponseInterface objects.

If the user is in async mode, they would get Promise objects or arrays of Promise objects. But how do they resolve them? Simple! Our library should have `InstagramAPI::resolvePromises()` which takes 1 argument: Either a single Promise, or an array of Promises. And the function simply loops through all array values recursively, checks if any is a Promise, and resolves them all. (There could be non-Promise values in the argument too, for things like endpoint functions that use `setAllowAsync(false)`).

The user could then do:

```
$ig = new InstagramAPI();
$ig->setAsync(true);
$data = $ig->resolvePromises([
    'userinfo1' => $ig->people->getUserInfoById(123),
    'userinfo2' => $ig->people->getUserInfoById(456),
    'wholeprofile' => $ig->people->getUserProfile(999),
    'whatever' => $ig->timeline->getSelfTimelineFeed(),
]);
$
```

It's then up to `resolvePromises` to resolve the array:

- The input may be a single value or an array. If a single value, just wrap it in an array `[$value]` to make the rest of the parsing code easier.
- It may be a multi-level array, since some "multi-action" endpoints (as you have proposed) return arrays of data. So the input array must be looped and resolved recursively.
- In fact, the user themselves may have made a multi-level array, with stuff like `['user1'=>['data1'=>endpoint,'data2'=>endpoint],'user2'=>...]`.
- As I said above, some of the array entries may _already_ be `ResponseInterface` objects due to those endpoints having been synchronous. So the `resolvePromises()` algorithm must _only_ look for `InstagramPromise` objects when resolving promises in the input array/value.
- Also, _every_ `InstagramPromise` object _MUST_ define a field that says _what_ endpoint they refer to (_MINUS_ any arguments in the endpoint URL, so ie `/media/{id}/info/` would be treated as `media/info` so that we're able to check if two Promises refer to the exact same API endpoint). The `resolvePromises()` function _must_ queue things properly, through multiple iterations of the array, to prevent multiple calls to the _exact same_ endpoint.
- Furthermore, the `resolvePromises()` function will need a _VERY_ advanced algorithm to prevent non-sense... such as getting an input array that wants the userInfo for user1 and the storyfeed for user2. Those endpoint calls refer to different users, so they should NOT run at the same time. One should resolve, and then the other should resolve, to look more like a real app (because the real app would NEVER instantly query different info about 2 different user-profiles simultaneously). Perhaps by having a "targetId" property on every `InstagramPromise` object, with a value such as `user::123` or `media::12382183281_123821`, and NEVER letting two `InstagramPromise` objects resolve simultaneously unless they have the EXACT SAME TARGET ID. That way, the queue will never resolve two unrelated objects. It will ensure that all simultaneous requests relate to the same user, or the same media item, etc. And together with the previous point (which prevents multiple requests to the same API endpoint simultaneously), it would be safe at dividing up all work.
- And, to complicate things even more: **Instagram's server implements throttling of requests.** They get _very_ angry if a client does rapid requests at a non-human rate/interaction behavior interval. So to avoid getting ThrottledException, it would need to divide up all of the work using a _very, VERY, VEEERY, VEEEEEEEEEEEERY_ advanced algorithm which tries to balance the workload and sleep() properly between various chunks of requests. The user may have given it a HUGE array of work to do, and it should be able to space it all out and handle it without triggering any ThrottledException (or at least retrying X times if it's throttled).

Okay, so that's it for the implementation...

But what would all of that achieve?

- Almost **NOTHING** can run simultaneously anyway, because we'll only allow simultaneous requests when they relate to the EXACT SAME `targetId` (to emulate the real app which never asks about multiple different profiles/media items/thread items/comments/etc at once), and since all simultaneous requests MUST CALL DIFFERENT ENDPOINTS. Most users will want multiple requests to the SAME endpoint (such as `media/info`), or about DIFFERENT users/media items, so their requests would run sequentially via the algorithm I described above. One user/media item at a time. One endpoint at a time. **So the end result of this rewrite, for most users, is: Almost NOTHING in the entire API runs simultaneously. Almost everything is STILL sequential.**
- We would need incredibly advanced code for the `resolvePromises()` function, and it still wouldn't be perfect. It would need so much work to ensure it properly divides up work among related endpoints, related `targetIds`, etc. And it would have MASSIVE problems with encountering throttling due to all the rapid API calls.
- All endpoint functions would have to be CAREFULLY rewritten to add the `targetId` and `endpointUrl` properties to the `InstagramPromise` they return. So that they can be properly queued.
- Instagram would still know that we aren't a real API app client, because we don't implement their very complex "graph" analytics collection/periodic data submission. And we won't do it. Because the app does _WAY TOO MUCH_ random data collection and processing for us to _EVER_ implement their "graph" API properly.

So, to sum it up: A huge waste of programming-time. Almost every request a user would make under this new system would still run synchronously (one at a time) due to the _necessary_ "safe resolution queue" algorithm. (Exactly like in the real app, which only does a single "unit of work" at a time). And we would only save about _half of a second_ in _some_ requests (ie "getUserProfile" which calls 4 endpoints); but we _didn't even need to_ make those functions anyway because Instagram doesn't care if we only call a single user-info endpoint (the one we need) instead of all four like the real app. And at the end of all that programming work, Instagram can _STILL_ detect our library, because we don't implement Graph.

So... it's up to you. If you do it, then do it as described here. But be prepared for about 3 months of your life being lost. ;-)

Would you like to join the project and work on some _worthwhile_ changes instead? Because your intelligence could definitely be better used for implementing _other_ stuff that's actually beneficial.

**Comment by SteveJobzniak** [comment]
^ I have just updated the post, in case you've already read it.

Closing this ticket.

**Comment by hamisor** [comment]
Hi @SteveJobzniak, thank your very much for your detailed analysis and suggestions on how to implement the asynchronous feature for this library. This is a very good project indeed. I'll have to read the core business logic of this project more thoroughly before I can contribute to it. It is a very good learning journey though even from the feedback you provided here. Really appreciated!

**Comment by SteveJobzniak** [comment]
@hamisor Hey, apologies for the long delay in replying (I was sick).

Anyway, you are always very welcome to join in the future. Brilliant minds are always in high demand! ;-)
